File: go/connect.go
package connect

var dirs = [][2]int{
	{0, -1},
	{0, 1},
	{-1, 0},
	{-1, 1},
	{1, -1},
	{1, 0},
}

func check(lines []string, q [][2]int, target byte, fn func(i, j int) bool) bool {
	visited := map[[2]int]bool{}

	for len(q) > 0 {
		pos := q[0]
		q = q[1:]

		if visited[pos] {
			continue
		}

		visited[pos] = true

		i, j := pos[0], pos[1]

		if fn(i, j) {
			return true
		}

		for _, d := range dirs {
			ni := i + d[0]
			nj := j + d[1]

			if ni >= 0 && ni < len(lines) && nj >= 0 && nj < len(lines[ni]) && lines[ni][nj] == target {
				q = append(q, [2]int{ni, nj})
			}
		}

	}

	return false
}

func checkPlayerO(lines []string) bool {
	q := [][2]int{}
	for i, c := range lines[0] {
		if c == 'O' {
			q = append(q, [2]int{0, i})
		}
	}

	return check(lines, q, 'O', func(i, j int) bool {
		return i == len(lines)-1
	})
}

func checkPlayerX(lines []string) bool {
	q := [][2]int{}
	for i := range lines {
		if lines[i][0] == 'X' {
			q = append(q, [2]int{i, 0})
		}
	}

	return check(lines, q, 'X', func(i, j int) bool {
		return j == len(lines[i])-1
	})
}

func ResultOf(lines []string) (string, error) {
	if checkPlayerO(lines) {
		return "O", nil
	}

	if checkPlayerX(lines) {
		return "X", nil
	}

	return "", nil
}

File: ocaml/connect.ml
open Base

type player = O | X

let check board target =
  let row = Array.length board
  and col = Array.length board.(0)
  and q = Queue.create ()
  and seen = Hash_set.Poly.create () in
  let is_win (i, j) =
    match target with
    | 'X' -> j = col - 1
    | 'O' -> i = row - 1
    | _ -> failwith "error"
  in
  let rec bfs () =
    match Queue.dequeue q with
    | None -> false
    | Some (i, j) ->
        if Hash_set.Poly.mem seen (i, j) then bfs ()
        else if is_win (i, j) then true
        else (
          Hash_set.Poly.add seen (i, j);
          [
            (i, j + 1);
            (i - 1, j);
            (i, j - 1);
            (i + 1, j);
            (i + 1, j - 1);
            (i - 1, j + 1);
          ]
          |> List.iter ~f:(fun (ni, nj) ->
                 if
                   ni >= 0
                   && ni < row
                   && nj >= 0
                   && nj < col
                   && Char.(board.(ni).(nj) = target)
                 then Queue.enqueue q (ni, nj));
          bfs ())
  in
  Char.(
    if target = 'X' then
      for i = 0 to row - 1 do
        if board.(i).(0) = target then Queue.enqueue q (i, 0)
      done
    else
      for j = 0 to col - 1 do
        if board.(0).(j) = target then Queue.enqueue q (0, j)
      done);
  bfs ()

let connect lst =
  let board =
    List.map lst ~f:(fun s ->
        String.to_list s
        |> List.filter ~f:(fun c -> Char.(c <> ' '))
        |> Array.of_list)
    |> Array.of_list
  in
  if check board 'X' then Some X else if check board 'O' then Some O else None
